File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minNode(self, node: Optional[TreeNode]):
        while node.left:
            node = node.left
        return node.val

    def deleteNode(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return None
        if val > root.val:
            root.right = self.deleteNode(root.right, val)
        elif val < root.val:
            root.left = self.deleteNode(root.left, val)
        else:
            if not root.left:
                return root.right
            elif not root.right:
                return root.left
            else:
                root.val = self.minNode(root.right)
                root.right = self.deleteNode(root.right, root.val)
        return root

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var minNode = function(curr) {
    while (curr.left) {
        curr = curr.left
    }
    return curr.val
}
var deleteNode = function(root, val) {
    if (!root) {
        return null
    }
    if (val > root.val) {
        root.right = deleteNode(root.right, val)
    }
    else if (val < root.val) {
        root.left = deleteNode(root.left, val)
    }
    else {
        if (!root.left) {
            return root.right
        }
        else if (!root.right) {
            return root.left
        }
        else {
            root.val = minNode(root.right)
            root.right = deleteNode(root.right, root.val)
        }
    }
    return root
};